spotlight-client/src/datastores/Metric.ts

```diff
@@ -0,0 +1,321 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+// =============================================================================
+
+import { ValuesType } from "utility-types";
+import { callMetricsApi, DataFile } from "../metricsApi/metricsClient";
+import { NamedEntity } from "./types";
+
+export enum MetricTypes {
+  SentencePopulationCurrent = "SentencePopulationCurrent",
+  SentenceTypesCurrent = "SentenceTypesCurrent",
+}
+
+type MetricsContent = NamedEntity & { methodology: string };
+
+type MetricFactoryOptionsBase = MetricsContent;
+
+type TotalIdentifier = "ALL";
+const TOTAL_KEY: TotalIdentifier = "ALL";
+
+type DemographicView = "total" | "race" | "gender" | "age";
+
+const DIMENSION_DATA_KEYS: Record<
+  Exclude<DemographicView, "total">,
+  keyof DemographicFields
+> = {
+  age: "ageBucket",
+  gender: "gender",
+  race: "raceOrEthnicity",
+};
+
+// TODO: should these be enums?
+type RaceIdentifier =
+  | TotalIdentifier
+  | "AMERICAN_INDIAN_ALASKAN_NATIVE"
+  | "ASIAN"
+  | "BLACK"
+  | "HISPANIC"
+  | "NATIVE_HAWAIIAN_PACIFIC_ISLANDER"
+  | "WHITE"
+  | "OTHER";
+type GenderIdentifier = TotalIdentifier | "FEMALE" | "MALE";
+type AgeIdentifier =
+  | TotalIdentifier
+  | "<25"
+  | "25-29"
+  | "30-34"
+  | "35-39"
+  | "40<";
+
+type BaseMetricFields = {
+  stateCode: string;
+};
+
+type DemographicFields = {
+  raceOrEthnicity: RaceIdentifier;
+  gender: GenderIdentifier;
+  ageBucket: AgeIdentifier;
+};
+
+type LocalityFields = {
+  locality: string;
+};
+
+type RawRecord = Record<string, unknown>;
+
+export type AnyMetric = Metric<RawRecord>;
+
+type PopulationBreakdownByLocationRecord = BaseMetricFields &
+  DemographicFields &
+  LocalityFields & {
+    population: number;
+  };
+
+// TODO: should these even be different types? why not just PopulationSnapshot?
+type PopulationSnapshot = Metric<PopulationBreakdownByLocationRecord>;
+
+type SentenceTypeByLocationRecord = BaseMetricFields &
+  DemographicFields &
+  LocalityFields & {
+    dualSentenceCount: number;
+    incarcerationCount: number;
+    probationCount: number;
+  };
+type SentenceTypesCurrent = Metric<SentenceTypeByLocationRecord>;
+
+type DataTransformer<RecordFormat> = (rawData: DataFile) => RecordFormat[];
+
+type InitParams<RecordFormat> = {
+  contentSource: MetricsContent;
+  dataTransformer: DataTransformer<RecordFormat>;
+  sourceFileName: string;
+  // NOTE: these fields can only be populated if the RecordFormat contains the requisite fields;
+  // however, they are not REQUIRED to be populated in those cases (e.g., population snapshots
+  // contain demographic fields but do not need to support demographic filters).
+  // I think this means it's possible to inadvertently break a filter by setting it to `undefined`
```

Comment:
yeah for sure. I think this limited implementation is a gesture in that direction — for example `defaultDemographicView` is restricted to the `DemographicView` strings ("total", "race", "gender", "age").

One option I thought of to address the broken filter issue (I pushed an example of it) is to always require that this value be set and then support an explicit "nofilter" value for when we want to show everything in the UI simultaneously rather than one dimension at a time